java-api/src/main/java/com/sap/cloud/security/x509/X509Parser.java

```diff
@@ -0,0 +1,132 @@
+package com.sap.cloud.security.x509;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+import java.util.Objects;
+
+public class X509Parser {
+	public static final Logger LOGGER = LoggerFactory.getLogger(X509Parser.class);
+	public static final String BEGIN_CERTIFICATE = "-----BEGIN CERTIFICATE-----";
+	public static final String END_CERTIFICATE = "-----END CERTIFICATE-----";
+
+	private X509Parser() {
+	}
+
+	/**
+	 * Generates X509 thumbprint from provided certificate as specified in JWT
+	 * Certificate Thumbprint Confirmation Method
+	 * 
+	 * @see <a href="https://tools.ietf.org/html/rfc8705#section-3.1"></a>
+	 * @param certificate
+	 *            Base64 encoded X509 certificate
+	 * @return Thumbprint of X509 certificate
+	 * @throws NoSuchAlgorithmException
+	 *             when a particular cryptographic algorithm is requested but is not
+	 *             available in the environment
+	 * @throws CertificateException
+	 *             is thrown if errors occur while decoding X509 certificate
+	 */
+	public static String getX509Thumbprint(String certificate) throws NoSuchAlgorithmException, CertificateException {
+		X509Certificate x509Certificate = decodeX509(
+				Objects.requireNonNull(certificate, "X509 certificate can't be null"));
+		MessageDigest sha256Digest = MessageDigest.getInstance("SHA-256");
+		byte[] hashedX509 = sha256Digest.digest(x509Certificate.getEncoded());
+		String base64UrlEncoding = Base64.getUrlEncoder().encodeToString(hashedX509);
+		LOGGER.debug("Base64-url encoded thumbprint: {}", base64UrlEncoding);
+		return normalizeThumbprint(base64UrlEncoding);
+	}
+
+	/**
+	 * Thumbprint generation alternative to {@link #getX509Thumbprint(String)}
+	 * without decoding into {@link X509Certificate}class
+	 *
+	 * @see <a href="https://tools.ietf.org/html/rfc8705#section-3.1"></a>
+	 * @param base64certificate
+	 *            certificate encoded in Base64, supports also PEM encoded
+	 *            certificates
+	 * @return Thumbprint of X509 certificate
+	 * @throws NoSuchAlgorithmException
+	 *             when a particular cryptographic algorithm is requested but is not
+	 *             available in the environment
+	 */
+	public static String getCertificateThumbprint(String base64certificate) throws NoSuchAlgorithmException {
```

Comment:
for which purpose this method is there? I think this one isn't used and needed?

Reply:
that's described in javadoc, one of the 2 methods can be removed based on the strategy what we store in the  securitycontext -> cert encoded in base64 or X509 cert object